Add management command that saves entity data, locations, and sources in a zip archive

smcalilly opened the pull request:
## Overview
See title.

First step for #807.

### Notes
We're repeating work with the data import, but this could lead us to refactor the download out of the original import script.

The new management command has side effects outside of make because it writes to files within the management command, so Make isn't able to pickup on those dependencies. Let me know if you have any ideas on how to handle the file writing within Make.

When I run this command, it creates an `import_docket.csv` file in the root directory. Is there a way to clean this up that I'm missing?

## Testing
- pull down branch
- run `docker-compose run --rm app make sfm_download.zip`

smcalilly commented on docket.mk:
```diff
@@ -1,4 +1,4 @@
-%_import : %.csv
+%_import : %.csv data
```
this could be the dependency for using the new data download recipe in the existing import script

smcalilly commented on sfm_pc/management/commands/download_data.py:
```diff
@@ -0,0 +1,222 @@
+            output_directory=sources_directory
+        )
+        
+        self._create_entity_files(
```
`_create_sources_file` and `_create_entity_files` are doing very similar things but slightly different (we're getting the source file based on the key value). i wavered back and forth over whether or not to separate this (first version had them separate), but decided to include them together. i thought instantiating the sheets service once made sense.

hancush commented on sfm_pc/management/commands/download_data.py:
```diff
@@ -0,0 +1,222 @@
+            self.style.SUCCESS('Downloaded locations file to {}'.format(target))
+        )
+    
+    def _build_location_downloader(self, location_doc_id):
+        drive_service = self._build_google_service(
+            scopes=['https://www.googleapis.com/auth/drive.readonly'],
+            service='drive',
+            version='v3'
+        )
+
+        request = drive_service.files().get_media(fileId=location_doc_id)
+        
+        location_buffer = io.BytesIO()
+        return MediaIoBaseDownload(location_buffer, request), location_buffer
```
I don't think you're getting much from breaking this out into its own method. Open to moving this code to the method that calls it?

smcalilly replied:
I originally had this code in the method caller but the caller was a big function with a lot of stuff going on, so I encapsulated the logic into this discrete function to make the caller more readable and organized — trying to follow the loose rule that functions should have no more than 4 lines (impossible but a good ideal to reach for), functions should accept input and produce output (the output here is a download object and a buffer, which are interdependent), and things (aka this function) should do one thing and one thing well. 

@hancush With that in mind, is it okay if I leave it there?

hancush commented on sfm_pc/management/commands/download_data.py:
```diff
@@ -0,0 +1,222 @@
+        sources_doc_id = kwargs['sources_doc_id']
+        country_code = kwargs['country_code'].rstrip()
+        
+        parent_directory = f'data'
```
What if this were configurable, with a default to `country_data`?

hancush commented on sfm_pc/management/commands/download_data.py:
```diff
@@ -0,0 +1,222 @@
+        country_code = kwargs['country_code'].rstrip()
+        
+        parent_directory = f'data'
+        country_directory = f'{parent_directory}/countries/{country_code}'
```
```suggestion
        country_subdirectory = f'{parent_directory}/countries/{country_code}'
```

hancush commented on sfm_pc/management/commands/download_data.py:
```diff
@@ -0,0 +1,222 @@
+        sources_doc_id = kwargs['sources_doc_id']
+        country_code = kwargs['country_code'].rstrip()
+        
+        parent_directory = f'data'
+        country_directory = f'{parent_directory}/countries/{country_code}'
```
It's a little confusing to define these here and pass them around. What do you think about storing the output directory and country code as class attributes, then creating the filepaths in the methods that actually create the files?

smcalilly replied:
see comment above

hancush reviewed (changes requested):
Looking good, thank, @smcalilly! Left a few thoughts inline.

smcalilly commented on sfm_pc/management/commands/download_data.py:
```diff
@@ -0,0 +1,222 @@
+    def _create_sources_file(self, sheets_service, document_id, output_directory):
+        target = f'{output_directory}/sources.csv'
+        
+        # No need to keep downloading the same data
+        if os.path.exists(target):
+            return
```
if i re-combine the two csv functions with the key lambda, and call this method, then it will skip in the case of a fresh import. so we need to be sure to add a way in make to remove these files after the import finishes.

smcalilly commented on sfm_pc/management/commands/download_data.py:
```diff
@@ -0,0 +1,222 @@
+        )
+        
+        for key, sheet in local_workbook.items():
+            if key == 'sources':
```
Good idea, I considered this but thought you'd be like what? Glad you're cool with that because that pattern can be useful (sparingly).

smcalilly commented on docket.mk:
```diff
@@ -1,4 +1,4 @@
-%_import : %.csv
+%_import : %.csv data
```
Oh cool, I was afraid it might be more work to plug in the local files. I'll see if I can use this existing option and rework if I need to. Do you want me to do this in this PR or another?

hancush replied:
This PR would be great!

smcalilly commented on sfm_pc/management/commands/download_data.py:
```diff
@@ -0,0 +1,222 @@
+        )
+
+    def handle(self, *args, **kwargs):
+        entity_doc_id = kwargs['entity_doc_id']
+        location_doc_id = kwargs['location_doc_id']
+        sources_doc_id = kwargs['sources_doc_id']
+        country_code = kwargs['country_code'].rstrip()
```
I agree that passing the arguments around can be a bit confusing. I think the latest version of this code makes more sense without setting them as class attributes, because the `_create_csv_files` caller is more generic with the `doc_id` (and so are the functions within that function that use that parameter). With the latest change, I don't know if it's necessary anymore to add them as attributes. Take a look and let me know if you agree?

Aside, I thought about adding them as attributes originally, but decided to pass as arguments because I understand class attributes to be more useful for state-like objects, where passing arguments is more pure function-like which makes it easier to understand the input for a function rather than having to grok the entire class. I sometimes find class attributes confusing. They can feel like they *come from above* and code makes more sense to me when arguments are passed directly into the function call, though admittedly it can get confusing when you pass the same argument to multiple levels of function callers (in which case that it might be better to add as an attribute). This isn't based on any principle, just my own style hunch and impression from code examples I've seen. There's a philosophical debate over "is this class a thing with these attributes" but this is work, not play!

hancush reviewed (commented):
Nice improvement, thanks @smcalilly. Can you figure out how to read from the downloaded files and update the import recipe? Then this is good to go.

smcalilly commented on sfm_pc/management/commands/import_google_doc.py:
```diff
@@ -420,26 +289,36 @@ def get_sheets_from_folder(self, folder):
-                # We could use csv.DictReader directly here, but instead use
-                # the format_dict_reader convenience method to test it
-                records = self.format_dict_reader(list(reader))
+            records = self.get_records_from_csv(path)
```
moved this logic into a function because we're calling it twice. since sources isn't living in each country's directory, we can include it in the `entities` tuple and instead have to call it by hand on line 303.

smcalilly replied:
i just realized a way we might could include the sources in the entities tuple — just go ahead and build up the filename in the tuple

smcalilly commented:
@hancush this is ready for review again.

i removed the google sheet logic from the import_google_doc mgmt command, so now the import process only loads data from a local directory. this required me to change the way we handle the persons_extra file since it's not always present in the data. i also had to change the file target for some of the test fixtures — hopefully that's self-explanatory.

also, i renamed the two commands but had to undo it because it treated the existing file as deleted and the diff showed the entire file. i thought that made it difficult to review, so i'll rename the files like this once this pr is approved:
- existing `import_google_doc.py` -> `load_country_data.py`
- new `download_country_data.py` -> `import_google_doc.py`

let me know if you have ideas for a different naming scheme.

hancush commented on sfm_pc/management/commands/import_google_doc.py:
```diff
@@ -121,22 +113,19 @@ class Command(BaseCommand):
-            dest='source_doc_id',
-            default='15lChqP4cKsjk8uUbUTaUA5gOUM09-t9YMIODORHw--8',
-            help='Import source data from specified Google Drive Document'
+            '--country_directory',
```
Can you standardize on directory or path for the country data and source file arg? Probably path, since it works for both directories and files.

hancush reviewed (changes requested):
Looking good, @smcalilly. Left a few comments inline.

Re: naming, I'd prefer to exorcise the `import_google_doc` name entirely. How about `download_country_data` and `import_country_data`?

Finally, can you think of any ways to test your new download management command? Likely don't want to hit a third-party service, but curious about your thoughts!

smcalilly commented on sfm_pc/management/commands/import_google_doc.py:
```diff
@@ -226,16 +195,10 @@ def handle(self, *args, **options):
-        else:
-            self.stdout.write('Loading data from Google Sheets...')
-            all_sheets = self.get_sheets_from_doc(options['doc_id'], options['source_doc_id'])
+        if (country_directory := options.get('country_directory').strip()):
```
Good question, I was just following the existing pattern. It checked if the variables existed to determine if it would load via google docs or local files. So now that is pointless.

smcalilly commented on docket.mk:
```diff
@@ -1,12 +1,27 @@
+sfm_download.zip : sfm_pc/management/commands/country_data
+	zip -r $@ $<
+
+sfm_pc/management/commands/country_data : import_docket.csv
```
I'm not sure I understand Make enough to get what you're saying, could you show me a code example @hancush? My understanding is that is already a target. Are you suggesting to do something like

```
country_data: sfm_pc/management/commands/country_data
```

so you can just run `make country_data` instead of the entire directory path?

smcalilly replied:
Or are you suggesting to make all of the individual country files a target?

hancush replied:
Ah, sorry, I misread your Makefile.

hancush commented on docket.mk:
```diff
@@ -1,12 +1,27 @@
-%_import : %.csv
+%_import : %.csv sfm_pc/management/commands/country_data sfm_download.zip
```
Do you need `sfm_download.zip` for the normal import to run? If not, you can omit it here and add as dependency of the target to upload the archive to S3, later.

smcalilly replied:
Correct, it's like that in my other branch where I've implemented the upload.

smcalilly commented:
@hancush I'm not sure a good way to test the google sheets api, but here are some options. Both would include writing a new test(s).

- The `google-api-python-client` is apparently just an API wrapper, so we could mock the requests like any other third-party api test.
- Add the api credentials file to our github action runner so it has the credentials and runs a test that requests the api. So whenever we run tests locally, it would presumably fail unless a developer had that file (which they'll most likely have). I don't know best practices with secrets in github actions, other than the fact that you can add them as environment variables (?). The api secrets are an entire file, would that work too?

hancush commented:
@smcalilly Yes, please

smcalilly commented:
@hancush this is ready to review again. I made requested changes and also changed a small bit of the readme about the import.

hancush commented on README.md:
```diff
@@ -93,15 +93,15 @@ If you are not on the keyring, enable the Google Sheets and Google Drive APIs in
     # Import entity data from default files
-    docker-compose run --rm app make import_google_docs
+    docker-compose run --rm app make import_country_data
 
     # Import entity data from specified files
-    docker-compose run --rm app python manage.py import_google_doc --doc_id ${SOME_ID} --source_doc_id ${SOME_ID} --location_doc_id ${SOME_ID}
+    docker-compose run --rm app python manage.py download_country_data --entity_doc_id ${SOME_ID} --sources_doc_id ${SOME_ID} --location_doc_id ${SOME_ID} --country_code ${SOME_CODE} --parent_directory ${DIRECTORY}
```
Is it possible to import without downloading, or should we instruct folks to always run download, then import?

smcalilly replied:
Ah, good catch, realizing this might be bad information — I'll rework. Do you think we need this "# Import entity data from specified files" instruction? That information exists in the "Importing the data" section. There's a little bit of overlap in information, but I see a little bit of value in keeping the section here for setting up local development while still having a separate "importing the data" section that gets at a lower level.

hancush replied:
I'll take redundancy for the sake of ease. I think I'd rework like:

```bash
# Download country data listed in fixtures/import_docket.csv
... download command ...

# Import downloaded country data
... import command ...
```

Thoughts?

smcalilly replied:
@hancush Agreed about the sake of ease. What about removing the individual steps and just adding the command for the import_docket_import make recipe, since that recipe calls both commands and would import the data for local development? (I think that would achieve that, right?) When I first started on this project, I found it kind of confusing to have the 

```
docker-compose run --rm app python manage.py import_google_doc --doc_id ${SOME_ID} --source_doc_id ${SOME_ID} --location_doc_id ${SOME_ID}
```

command spelled out as an instruction rather than an example, especially since it's explained later in the doc. There are a bunch of different ways to import the data in this readme, so it's helpful for me to have an explicit, one way instruction for setting up local development. And then later include more granular stuff (examples and different ways to do it). 

I think it's a question of what this section should contain. It's called "Load data via management commands" and the readme in the master branch says to run `import_google_docs` make recipe but I don't see that anymore. Do we want developers to run the commands for each entity by hand, or run individual make recipes, or an instruction with an automated make process so they're setup to develop? And do we expect a dev to download/import all of the data or just work with a little bit of it?

hancush replied:
Good thought, @smcalilly. Let's do `make import_docket_import`, and point to docs further down on loading a subset of data, if they want to.

smcalilly replied:
If we want the local dev setup to only include a snippet of the data then I think your suggestion makes the most sense.

hancush replied:
@smcalilly I endow you with the power to make a decision here, haha.

smcalilly replied:
I think you had it right after seeing that section as a whole, changed here 4336c21c

hancush reviewed (commented):
Thanks so much, @smcalilly, especially for thinking to update the README. One small README edit inline, then this can come in.
